### DISCUSSION TITLE
Multiple Replicates best practices?

### DISCUSSION THREAD

**Discussion by vdlmrc** [chemprop/chemprop#1134] [Q&A] [answered]
Hi Chemprop community,

I am still learning chemprop (version 2.1.0), and read the few discussions about the decision to not implement cross validation and use mutiple replicates instead.

I am working on a classifier, and I look at MCC and a bunch of other metrics, but let's focus on MCC for this discussion.
I ran 10 replicates on one architecture and found the MCC to be highly variable. **Min=-0.01**, **Max=0.51**, **Average=0.35**.
I tested a couple of hyperparameters combinations and always found significant variability. Not always as dramatic as above, but still concerning.

**I would like to know what is the best practice with multiple replicates:**
-  How do you conduct an hyperparamter optimization? 
    - Do you run replicates and elect the best combination based on highest average and lowest standard deviation of the MCC (or your metric of choice)?
    - If you add 10 replicates for each hyperparamters the computing time/power skyrockets. And even 10 replicates seems quite low with this kind of variability. I would go for 100 replicates if computing power wasn't an issue. Any guidance on that?
- Once you have your best hyperparameter combination, how do you proceed?
    - Run 100 repliactes and pick the model closest to the average MCC value, or cherry pick the highest MCC, considering the way we chosed our hyperparameters would prevent overfitting?
- Any other advice to make the most of the mutliple replicates strategy?

Many thanks!

Marc

**Comment by JacksonBurns** [comment, marked as answer]
>       * Do you run replicates and elect the best combination based on highest average and lowest standard deviation of the MCC (or your metric of choice)?

The ideal set of hyperparameters will change based on the choice of validation set for any given replicate, which will change each time. Ideally, they parameters won't change _too much_ i.e. the best architecture should be mostly consistent. If this isn't the case, it means your model is probably overfitting - you may want to consider something like adding dropout.

>       * If you add 10 replicates for each hyperparamters the computing time/power skyrockets. And even 10 replicates seems quite low with this kind of variability. I would go for 100 replicates if computing power wasn't an issue. Any guidance on that?

Huge variability in performance can sometimes reflect issues with the dataset. You may want to consider balanced sampling, augmentation, etc.

>       * Run 100 repliactes and pick the model closest to the average MCC value, or cherry pick the highest MCC, considering the way we chosed our hyperparameters would prevent overfitting?

I would recommend training some reasonable number of replicates and using the ensemble of models together as the predictor - that way effectively all of the training data has been used.
